subprojects/architecture-test/src/test/java/org/gradle/architecture/test/ProviderMigrationArchitectureTest.java

```diff
@@ -0,0 +1,133 @@
+        @Override
+        public boolean apply(JavaMethod input) {
+            PropertyAccessorType accessorType = PropertyAccessorType.fromName(input.getName());
+            return accessorType == PropertyAccessorType.GET_GETTER || accessorType == PropertyAccessorType.IS_GETTER;
+        }
+    };
+
+    private static final DescribedPredicate<JavaMethod> haveSetters = new DescribedPredicate<JavaMethod>("have mutable property") {
+        @Override
+        public boolean apply(JavaMethod input) {
+            PropertyAccessorType accessorType = PropertyAccessorType.fromName(input.getName());
+            String propertyNameFromGetter = accessorType.propertyNameFor(input.getName());
+            return input.getOwner().getAllMethods().stream()
+                .filter(method -> PropertyAccessorType.fromName(method.getName()) == PropertyAccessorType.SETTER)
+                .anyMatch(method -> PropertyAccessorType.SETTER.propertyNameFor(method.getName()).equals(propertyNameFromGetter));
+        }
+    };
+
+    private static final DescribedPredicate<JavaClass> class_with_any_mutable_property = new DescribedPredicate<JavaClass>("class with any mutable property") {
+        @Override
+        public boolean apply(JavaClass input) {
+            return input.getAllMethods().stream()
+                .filter(getters::apply)
+                .anyMatch(haveSetters::apply);
+        }
+    };
+
+    private static final DescribedPredicate<JavaMethod> mutable_public_API_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(not(declaredIn(assignableTo(Task.class))))
+        .and(not(declaredIn(StartParameter.class)))
+        .and(not(declaredIn(Configuration.class)))
+        .and(not(declaredIn(FileCollection.class)))
+        .and(not(declaredIn(ConfigurableFileCollection.class)))
+        .and(are(declaredIn(class_with_any_mutable_property)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .as("mutable public API properties");
+
+    @SuppressWarnings("deprecation")
+    private static final DescribedPredicate<JavaMethod> task_properties = ArchPredicates.<JavaMethod>are(public_api_methods)
+        .and(declaredIn(assignableTo(Task.class)))
+        .and(are(getters))
+        .and(not(annotatedWith(Inject.class)))
+        .and(not(declaredIn(Task.class)))
+        .and(not(declaredIn(DefaultTask.class)))
+        .and(not(declaredIn(org.gradle.api.internal.AbstractTask.class)))
+        .as("task properties");
+
+    @ArchTest
+    public static final ArchRule mutable_public_api_properties_should_be_providers = freeze(methods()
+        .that(are(mutable_public_API_properties))
+        .and().doNotHaveRawReturnType(TextResource.class)
+        .and().doNotHaveRawReturnType(assignableTo(FileCollection.class))
+        .should().haveRawReturnType(assignableTo(Provider.class)));
```

Comment:
Why did we decide on this being a `Provider` instead of a `Property`? Since they are mutable, I guess they should be `Property`.